request.go

```diff
@@ -3,376 +3,145 @@ package cmds
-		if found {
-			return &cmdkit.OptionValue{Value: val, ValueFound: found, Def: option}
-		}
+// NewRequest returns a request initialized with given arguments
+// An non-nil error will be returned if the provided option values are invalid
+func NewRequest(ctx context.Context, path []string, opts cmdkit.OptMap, args []string, file files.File, root *Command) (*Request, error) {
+	if opts == nil {
+		opts = make(cmdkit.OptMap)
 	}
 
-	return &cmdkit.OptionValue{Value: option.Default(), ValueFound: false, Def: option}
-}
-
-// Options returns a copy of the option map
-func (r *request) Options() cmdkit.OptMap {
-	output := make(cmdkit.OptMap)
-	for k, v := range r.options {
-		output[k] = v
+	cmd, err := root.Get(path)
+	if err != nil {
+		return nil, err
 	}
-	return output
-}
 
-func (r *request) SetRootContext(ctx context.Context) error {
-	ctx, err := getContext(ctx, r)
-	if err != nil {
-		return err
+	req := &Request{
+		Path:      path,
+		Options:   opts,
+		Arguments: args,
+		Files:     file,
+		Root:      root,
+		Command:   cmd,
+		Context:   ctx,
 	}
 
-	r.rctx = ctx
-	return nil
+	return req, req.convertOptions(root)
 }
 
-// SetOption sets the value of the option for given name.
-func (r *request) SetOption(name string, val interface{}) {
-	// find the option with the specified name
-	option, found := r.optionDefs[name]
-	if !found {
-		return
+// BodyArgs returns a scanner that returns arguments passed in the body as tokens.
+func (req *Request) BodyArgs() (*bufio.Scanner, error) {
+	if len(req.Arguments) >= len(req.Command.Arguments) {
+		return nil, fmt.Errorf("all arguments covered by positional arguments")
 	}
 
-	// try all the possible names, if we already have a value then set over it
-	for _, n := range option.Names() {
-		_, found := r.options[n]
-		if found {
-			r.options[n] = val
-			return
-		}
+	if req.Files == nil {
+		return nil, fmt.Errorf("expected more arguments from stdin")
 	}
 
-	r.options[name] = val
-}
-
-// SetOptions sets the option values, unsetting any values that were previously set
-func (r *request) SetOptions(opts cmdkit.OptMap) error {
-	r.options = opts
-	return r.ConvertOptions()
-}
-
-func (r *request) StringArguments() []string {
-	return r.arguments
-}
-
-// Arguments returns the arguments slice
-func (r *request) Arguments() []string {
-	if r.haveVarArgsFromStdin() {
-		err := r.VarArgs(func(s string) error {
-			r.arguments = append(r.arguments, s)
-			return nil
-		})
-		if err != nil && err != io.EOF {
-			log.Error(err)
-		}
+	fi, err := req.Files.NextFile()
+	if err != nil {
+		return nil, err
 	}
 
-	return r.arguments
-}
-
-func (r *request) SetArguments(args []string) {
-	r.arguments = args
-}
-
-func (r *request) Files() files.File {
-	return r.files
+	return bufio.NewScanner(fi), nil
 }
 
-func (r *request) SetFiles(f files.File) {
-	r.files = f
+func IsAllArgsAlreadyCovered(err error) bool {
+	return err != nil && err.Error() == "all arguments covered by positional arguments"
```

Comment:
oooooo, can we try to avoid checking errors by their string?

Reply:
yeah, that was hacky. using type assertions on the error now.